File: example/matrix_float.java
/************************************************************
 @Author: HoodUSSEnterprise
 @Date: 2026-06-28
 @LastEditors: HoodUSSEnterprise
 @LastEditTime: 2026-06-28
 @FilePath: \asm_matrix_benchmark\example\matrix_float.java
 @Description: example of matrix float Java
*************************************************************/

public class matrix_float
{
    public static void main(String[] args)
    {
        float[] data = {1, 2, 3, 4};
        MatrixFloat v1 = new MatrixFloat(data, 2, 2);
        MatrixFloat v2 = new MatrixFloat(data, 1, 4);
        MatrixFloat v3 = new MatrixFloat(data, 2, 2);

        // add matrix example
        System.out.println("---------------------------------------add matrix---------------------------------------");
        MatrixFloat v = v1.add(v2);
        MatrixFloat m = v1.add(v3);
        if (v != null) v.print(); else System.out.println("add failed (expected)");
        if (m != null) m.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // sub matrix example
        System.out.println("---------------------------------------sub matrix---------------------------------------");
        v = v1.sub(v2);
        m = v1.sub(v3);
        if (v != null) v.print(); else System.out.println("sub failed (expected)");
        if (m != null) m.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // mul matrix example
        System.out.println("---------------------------------------mul matrix---------------------------------------");
        float[] mulData = {1, 2, 3};
        MatrixFloat mul1 = new MatrixFloat(mulData, 3, 1);
        MatrixFloat mul2 = new MatrixFloat(mulData, 1, 3);
        m = mul1.mul(mul2);
        if (m != null) m.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // scale matrix example
        System.out.println("--------------------------------------scale matrix--------------------------------------");
        float[] scaleData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        MatrixFloat scaleM = new MatrixFloat(scaleData, 2, 5);
        scaleM.print();
        m = scaleM.scale(2);
        if (m != null) m.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // cat matrix example
        System.out.println("---------------------------------------cat matrix---------------------------------------");
        float[] catData1 = {1, 2, 3};
        float[] catData2 = {4, 5, 6, 7, 8, 9};
        MatrixFloat catMatrix1 = new MatrixFloat(catData1, 1, 3);
        MatrixFloat catMatrix2 = new MatrixFloat(catData2, 2, 3);
        m = catMatrix1.cat(catMatrix2, 0);
        if (m != null) m.print();
        MatrixFloat catMatrix3 = new MatrixFloat(catData1, 3, 1);
        MatrixFloat catMatrix4 = new MatrixFloat(catData2, 3, 2);
        m = catMatrix3.cat(catMatrix4, 1);
        if (m != null) m.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // find matrix example
        System.out.println("--------------------------------------find matrix---------------------------------------");
        float[] findData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        MatrixFloat findMatrix = new MatrixFloat(findData, 5, 2);
        findMatrix.print();
        float findElem = 5;
        int[] pos = findMatrix.findElem(findElem);
        if (pos != null)
        {
            System.out.println("find elem");
            System.out.println("Position is (" + pos[0] + ", " + pos[1] + ")");
        }
        else
        {
            System.out.println("No find elem : " + findElem);
        }
        findElem = 11;
        pos = findMatrix.findElem(findElem);
        if (pos != null)
        {
            System.out.println("find elem");
            System.out.println("Position is (" + pos[0] + ", " + pos[1] + ")");
        }
        else
        {
            System.out.println("No find elem : " + findElem);
        }
        System.out.println("----------------------------------------------------------------------------------------");

        // replace matrix example
        System.out.println("-------------------------------------replace matrix-------------------------------------");
        float[] replaceData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        MatrixFloat replaceMatrix = new MatrixFloat(replaceData, 5, 2);
        int replaceX = 10, replaceY = 2;
        if (replaceMatrix.replaceByCoord(replaceX, replaceY, 10))
        {
            replaceMatrix.print();
        }
        else
        {
            System.out.println("(" + replaceX + ", " + replaceY + ") has out of index range");
        }
        replaceX = 0; replaceY = 0;
        if (replaceMatrix.replaceByCoord(replaceX, replaceY, 10))
        {
            replaceMatrix.print();
        }
        else
        {
            System.out.println("(" + replaceX + ", " + replaceY + ") has out of index range");
        }
        float oldDataF = 7;
        float newDataF = 10;
        if (replaceMatrix.replaceByValue(oldDataF, newDataF))
        {
            replaceMatrix.print();
        }
        else
        {
            System.out.println(oldDataF + " has not in matrix");
        }
        oldDataF = 20;
        if (replaceMatrix.replaceByValue(oldDataF, newDataF))
        {
            replaceMatrix.print();
        }
        else
        {
            System.out.println(oldDataF + " has not in matrix");
        }
        System.out.println("----------------------------------------------------------------------------------------");

        // transpose matrix example
        System.out.println("------------------------------------transpose matrix------------------------------------");
        float[] transposeData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        MatrixFloat rawMatrix = new MatrixFloat(transposeData, 5, 2);
        rawMatrix.print();
        MatrixFloat transposeMatrix = rawMatrix.transpose();
        if (transposeMatrix != null) transposeMatrix.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // special matrix example
        System.out.println("-------------------------------------special matrix-------------------------------------");
        MatrixFloat identity = MatrixFloat.identity(6);
        if (identity != null) identity.print();
        float[] diagData = {1, 2, 3, 4, 5};
        MatrixFloat diag = MatrixFloat.diag(diagData);
        if (diag != null) diag.print();
        MatrixFloat eye = MatrixFloat.eye(3, 4);
        if (eye != null) eye.print();
        MatrixFloat zero = MatrixFloat.zeros(10, 10);
        if (zero != null) zero.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // rank and trace example
        System.out.println("-------------------------------------rank and trace-------------------------------------");
        float[] rankData = {1, 2, 3, 4, 5, 6, 7, 8, 9};
        MatrixFloat rankMatrix = new MatrixFloat(rankData, 3, 3);
        int rank = rankMatrix.rank();
        if (rank != -1)
        {
            rankMatrix.print();
            System.out.println("matrix rank = " + rank);
        }
        MatrixFloat traceMatrix = new MatrixFloat(rankData, 3, 3);
        float trace = traceMatrix.trace();
        System.out.println("matrix trace = " + trace);
        System.out.println("----------------------------------------------------------------------------------------");

        // compare matrix example
        System.out.println("-------------------------------------compare matrix-------------------------------------");
        float[] compareData1 = {1, 2, 3, 4, 5, 6};
        float[] compareData2 = {6, 5, 4, 3, 2, 1};
        MatrixFloat compareMatrix1 = new MatrixFloat(compareData1, 2, 3);
        MatrixFloat compareMatrix2 = new MatrixFloat(compareData1, 2, 3);
        MatrixFloat compareMatrix3 = new MatrixFloat(compareData2, 2, 3);
        MatrixFloat compareMatrix4 = new MatrixFloat(compareData1, 3, 2);
        if (compareMatrix1.isEqual(compareMatrix2))
            System.out.println("Two matrix equals\n");
        else
            System.out.println("Two matrix not equals\n");
        if (compareMatrix1.isEqual(compareMatrix3))
            System.out.println("Two matrix equals\n");
        else
            System.out.println("Two matrix not equals\n");
        if (compareMatrix1.isEqual(compareMatrix4))
            System.out.println("Two matrix equals\n");
        else
            System.out.println("Two matrix not equals\n");
        System.out.println("----------------------------------------------------------------------------------------");

        // invertible matrix example
        System.out.println("-------------------------------------invertible matrix-------------------------------------");
        float[] invData = {1, 2, 3, 0, 4, 5, 1, 0, 6};
        MatrixFloat matrixOrigin = new MatrixFloat(invData, 3, 3);
        MatrixFloat matrixInv = MatrixFloat.inv(matrixOrigin);
        if (matrixInv != null) matrixInv.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // leading minors example
        System.out.println("-------------------------------------leading minors-------------------------------------");
        float[] leadingData = {1, 2, 3, 4, 5, 6, 7, 8, 9};
        MatrixFloat leadingMatrix = new MatrixFloat(leadingData, 3, 3);
        MatrixFloat[] leadingMinors = leadingMatrix.getLeadingMinors();
        if (leadingMinors != null)
        {
            for (int i = 0; i < leadingMinors.length; i++)
            {
                leadingMinors[i].print();
            }
        }
        System.out.println("----------------------------------------------------------------------------------------");

        // lu decomposition example
        System.out.println("-------------------------------------lu decomposition-------------------------------------");
        float[] luData = {1, 2, 3, 0, 4, 5, 1, 0, 6};
        MatrixFloat luMatrix = new MatrixFloat(luData, 3, 3);
        LUResult luRes = luMatrix.luDecomposition();
        if (luRes != null)
        {
            luRes.L.print();
            luRes.U.print();
        }
        System.out.println("----------------------------------------------------------------------------------------");

        // extract matrix example
        System.out.println("-------------------------------------extract matrix--------------------------------------");
        float[] extractData = {1, 2, 3, 4, 5, 6};
        MatrixFloat extractMatrix = new MatrixFloat(extractData, 2, 3);
        extractMatrix.print();
        MatrixFloat extractRow = extractMatrix.extractRow(0);
        if (extractRow != null) extractRow.print();
        MatrixFloat extractCol = extractMatrix.extractCol(1);
        if (extractCol != null) extractCol.print();
        System.out.println("extract diag: ");
        float[] extractDiag = extractMatrix.extractDiag();
        if (extractDiag != null)
        {
            for (int i = 0; i < extractDiag.length; i++)
            {
                System.out.print(extractDiag[i] + " ");
            }
            System.out.println();
        }
        System.out.println("----------------------------------------------------------------------------------------");

        // random matrix example
        System.out.println("-------------------------------------random matrix---------------------------------------");
        MatrixFloat randMatrix1 = MatrixFloat.random(3, 4, null);
        if (randMatrix1 != null) randMatrix1.print();
        float[] randRange = {5, 15};
        MatrixFloat randMatrix2 = MatrixFloat.random(2, 5, randRange);
        if (randMatrix2 != null) randMatrix2.print();
        float[] randRange2 = {20, 10};
        MatrixFloat randMatrix3 = MatrixFloat.random(4, 3, randRange2);
        if (randMatrix3 != null) randMatrix3.print();
        System.out.println("----------------------------------------------------------------------------------------");

        // determinant example
        System.out.println("-------------------------------------determinant-----------------------------------------");
        float[] detData = {1, 2, 3, 4};
        MatrixFloat detMatrix = new MatrixFloat(detData, 2, 2);
        double det = detMatrix.determinant();
        System.out.println("determinant = " + det);
        System.out.println("----------------------------------------------------------------------------------------");
    }
}

File: example/matrix_float.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 10:32:52
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-29 13:54:37
@FilePath: \asm_matrix_benchmark\example\matrix_float.cpp
@Description: example of matrix float
*************************************************************/

#include "matrix_float_cpp.h"
#include <cstdio>

int main()
{
    float data[4] = {1, 2, 3, 4};
    MatrixFloat v1(data, 2, 2);
    MatrixFloat v2(data, 1, 4);
    MatrixFloat v3(data, 2, 2);

    // add matrix example
    puts("---------------------------------------add matrix---------------------------------------");
    MatrixFloat m = v1 + v3;
    std::cout << m;
    puts("----------------------------------------------------------------------------------------");

    puts("---------------------------------------sub matrix---------------------------------------");
    m = v1 - v3;
    std::cout << m;
    puts("----------------------------------------------------------------------------------------");

    puts("---------------------------------------mul matrix---------------------------------------");
    float mul_data[3] = {1, 2, 3};
    MatrixFloat mul1(mul_data, 3, 1);
    MatrixFloat mul2(mul_data, 1, 3);
    m = mul1 * mul2;
    std::cout << m;
    puts("----------------------------------------------------------------------------------------");

    puts("--------------------------------------scale matrix--------------------------------------");
    float scale_data[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    MatrixFloat scale(scale_data, 2, 5);
    std::cout << scale;
    m = scale * 2;
    std::cout << m;
    puts("----------------------------------------------------------------------------------------");

    puts("---------------------------------------cat matrix---------------------------------------");
    float cat_data1[3] = {1, 2, 3};
    float cat_data2[6] = {4, 5, 6, 7, 8, 9};
    MatrixFloat cat1(cat_data1, 1, 3);
    MatrixFloat cat2(cat_data2, 2, 3);
    m = cat1.cat(cat2, 0);
    std::cout << m;
    puts("----------------------------------------------------------------------------------------");

    puts("--------------------------------------find matrix---------------------------------------");
    float find_data[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    MatrixFloat find_m(find_data, 5, 2);
    std::cout << find_m;
    auto p = find_m.find_elem(5);
    if (p.first != -1)
    {
        printf("find elem 5 at (%d, %d)\n", p.first, p.second);
    }
    else
    {
        printf("No find elem 5\n");
    }
    p = find_m.find_elem(11);
    if (p.first != -1)
    {
        printf("find elem 11 at (%d, %d)\n", p.first, p.second);
    }
    else
    {
        printf("No find elem 11\n");
    }
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------replace matrix-------------------------------------");
    float replace_data[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    MatrixFloat replace_m(replace_data, 5, 2);
    replace_m.replace_pos(0, 0, 99);
    std::cout << replace_m;
    replace_m.replace_elem(7, 77);
    std::cout << replace_m;
    puts("----------------------------------------------------------------------------------------");

    puts("------------------------------------transpose matrix------------------------------------");
    float transpose_data[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    MatrixFloat raw(transpose_data, 5, 2);
    std::cout << raw;
    MatrixFloat t = raw.transpose();
    std::cout << t;
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------special matrix-------------------------------------");
    MatrixFloat I = MatrixFloat::identity(4);
    std::cout << I;
    MatrixFloat Z = MatrixFloat::zeros(3, 4);
    std::cout << Z;
    MatrixFloat O = MatrixFloat::ones(2, 3);
    std::cout << O;
    MatrixFloat R = MatrixFloat::random(3, 3, 0, 10);
    std::cout << R;
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------rank and trace-------------------------------------");
    float rt_data[9] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    MatrixFloat rt_m(rt_data, 3, 3);
    std::cout << rt_m;
    printf("matrix rank = %zu\n", rt_m.rank());
    std::cout << "matrix trace = " << rt_m.trace() << std::endl;
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------compare matrix-------------------------------------");
    float comp1_data[6] = {1, 2, 3, 4, 5, 6};
    float comp2_data[6] = {6, 5, 4, 3, 2, 1};
    MatrixFloat cmp1(comp1_data, 2, 3);
    MatrixFloat cmp2(comp1_data, 2, 3);
    MatrixFloat cmp3(comp2_data, 2, 3);
    std::cout << (cmp1 == cmp2 ? "Two matrix equals\n" : "Two matrix not equals\n");
    std::cout << (cmp1 == cmp3 ? "Two matrix equals\n" : "Two matrix not equals\n");
    std::cout << (cmp1 != cmp3 ? "Two matrix not equals\n" : "Two matrix equals\n");
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------invertible matrix-------------------------------------");
    float inv_matrix[9] = {1, 2, 3, 0, 4, 5, 1, 0, 6};
    MatrixFloat matrix_orgin(inv_matrix, 3, 3);
    MatrixDouble matrix_inv(3, 3);
    matrix_inv = matrix_orgin.inv_matrix();
    std::cout << matrix_inv;
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------leading minors-------------------------------------");
    float lead_data[9] = {1, 2, 3, 4, 5, 6, 7, 8, 10};
    MatrixFloat lead_m(lead_data, 3, 3);
    std::cout << lead_m;
    auto minors = lead_m.leading_minors();
    printf("leading minors count = %zu\n", minors.size());
    for (size_t i = 0; i < minors.size(); i++)
    {
        printf("--- order %zu ---\n", i + 1);
        std::cout << minors[i];
    }
    puts("----------------------------------------------------------------------------------------");

    puts("-------------------------------------extract submatrix----------------------------------");
    float ext_data[12] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
    MatrixFloat ext_m(ext_data, 3, 4);
    std::cout << ext_m;
    MatrixFloat row = ext_m.extract_row(1);
    std::cout << row;
    MatrixFloat col = ext_m.extract_col(2);
    std::cout << col;
    MatrixFloat block = ext_m.extract_submatrix(0, 2, 1, 3);
    std::cout << block;
    puts("----------------------------------------------------------------------------------------");

    return 0;
}
